thetis/configuration.py

```diff
@@ -213,6 +213,31 @@ def validate(self, obj, value):
         self.error(obj, value)
 
 
+class CoupledTracerImplicitness(Dict):
+    info_text = 'a dictionary of implicness values for CoupledTracerPicard'
+
+    def __init__(self, default_value=Undefined, bounds=None, **kwargs):
+        self.minval = bounds[0]
+        self.maxval = bounds[1]
+        super(CoupledTracerImplicitness, self).__init__(default_value=default_value, **kwargs)
+
+    def validate(self, obj, value):
+        if (isinstance(value, dict)
+                and all(self.minval <= theta <= self.maxval for key, theta in value.items())):
+            return value
+        self.error(obj, value)
+
+
+class CoupledTracerSemiImplicitness(Dict):
+    info_text = 'a dictionary of bools toggling semi-implicitness for CoupledTracerPicard'
+
+    def validate(self, obj, value):
+        if (isinstance(value, dict)
+                and all(isinstance(semi_implicit, bool) for key, semi_implicit in value.items())):
+            return value
+        self.error(obj, value)
+
+
```

Comment:
I think this is way too specific for a new Traitlet type. A Traitlet type here should be a type that implements a type-checking validate method, so it should only be as specific as required to describe what sort of data is allowed. Here you seem to be implementing a `BoundedFloatDict` and a `BoolDict`.

Note however that traitlets allows you to compose the `Dict` type with a specified trait:
```
class MyOptions(HasFrozenTraits):
    x = Dict(value_trait=BoundedFloat(bounds=[0,1]), default_value={})
    y = CoupledTracerImplicitness(bounds=[0,1])
```
Both work as expected, but the former gives a slightly more useful error message:
```
z = MyOptions()
z.x = {'a': 1.5}
...
TraitError: Values of the 'x' trait of a MyOptions instance must be a float between 0 and 1, but a value of 1.5 <class 'float'> was specified.

z.y = {'b': -0.5}

TraitError: The 'y' trait of a MyOptions instance expected a dict, not the dict {'b': -0.5}
```

However there is a more fundamental problem with both approaches: the values only get type checked when assigning a whole new dictionary to the "bounded float dict" attribute, *not* when the dictionary is subsequently modified:
```
z = MyOptions()
z.x['a'] = 1.5
z.y['b'] = -0.5
```

Finally I'm not sure (up for discussion really) whether we want to go the route of having a series of dictionaries all with the same keys, for all the various options we want to set per field as the number of these options is inevitably going to increase. To me it would make more sense to have a hierarchy, where you only have one dictionary whose values are entire `FrozenHasTraits` objects with all the options for every field grouped together at that level. So something like:
```
class TracerFieldOptions(FrozenHasTraits):
    theta = BoundedFloat(bounds=[0,1])
    semi_implicit = Bool()

class MyOptions:
    tracer_options = Dict(value_trait=Instance(TracerFieldOptions), default_value={})
```
Then the user would
```
tracer_options_a = TracerFieldOptions()
tracer_options_a.theta = 0.5
tracer_options_a.semi_implicit = True
tracer_options_b = TracerFieldOptions()
tracer_options_b.theta = 0.5
tracer_options_b.semi_implicit = True
options = MyOptions() # this would be done by Thetis of course
options.tracer_options = {'a': tracer_options_a, 'b': tracer_options_b}
```
This has the advantage that type checking happens not only when the `TracerFieldOptions` are first filled in, but if the options are subsequently changed:
```
options.tracer_options['a'].theta = 1.5
```
gives the appropriate error. The user can however still shoot themselves in the foot  (in the same way as before) by doing:
```
options.tracer_options['a'] = 'foo' # or anything that's not a TracerFieldOptions instance
```
A disadvantage is that it exposes the user to this `TracerFieldOptions` class that the user has to instantiate so it adds some complexity...

I would really like to hear other people's opinions as the same issue applies to specifying multiple turbine farms, where we have a dictionary that maps region ids (2d gmsh physical surface ids) to options for the different farms. Soon to be expanded with a lot more options.

Reply:
Great point, using separate options seems much nicer, indeed. I get the point about user friendliness, though.

I guess in this case, the `ModelOptions2d.add_tracer_2d` interface could pass some of this through, but then we might end up with lots of different kwargs to account for all the timesteppers.

Reply:
Good point, Stephan! I agree that option hierarchy is probably the right way to go. As Joe says `add_tracer_2d` could indeed be used to reduce boilerplate e.g. by populating the parameters with default values. In passing I mention that we do currently have to use quite a bit of `hasattr` checks to sniff whether certain option exists which isn't very nice. I think it's the only downside of having more dynamic options structure.

Reply:
Ah yes, forgot we already have `add_tracer_2d` indeed. So maybe `tracer_field_options` (or however we'd call it) should be some sort of frozen dict that the user can't add entries to other than through calling add_tracer_2d, so like tracer_metadata but with the values being a TracerFieldOptions instance in which all the options are set for each field including name/shortname/filename. So the TracerFieldOptions instance gets created by add_tracer_2d (so the creation isn't exposed to the user) and added as an entry in the `tracer_field_options` dictionary. That way we enforce that the user sets the minimum for a tracer field to be functional, but any further options like e.g. `source`, take a default and can be set as attributes of the TracerFieldOptions instance: `options.tracer_field_options['foo'].source = Function(...)`

On the topic (slightly off) of `hasattr` - I think this is mainly an issue of feeding in the options at the right level to the corresponding class in the object hierarchy. In particular looking at the many `hasattr`s in the various `get_..._timestepper` methods of the solver objects - it probably makes more sense to just pass in the entire `timestepper_options` (a TimeStepperOptions (or subclass) instance) to the time integrator, rather than unpacking it via various keyword arguments.

Reply:
1. Sounds like a good solution to me.
2. I agree that would neaten things up. That way the `get_..._timestepper` methods just serve to feed in the appropriate parameters for the equation, rather than the timestepper (and therefore should probably be renamed).

Thanks both for the feedback, starting to think this will be easier to implement on a new branch, so will do that.

Reply:
[This ](https://github.com/thetisproject/thetis/pull/247) should do the trick for reducing boilerplate and standardising args/kwargs for the timesteppers.